### ISSUE TITLE
`ess-eval-buffer` hangs Emacs during evaluation with certain R code on macOS

### ISSUE THREAD

**Issue by ywwry66** [emacs-ess/ESS#1096]
Environments:
- macOS Big Sur/Catalina
- Emacs Mac Port and also Emacs development version (installed with brew install emacs --head)
- ess-20210113.915 and also older versions
- R version 4.0.3

I found this when investigating an issue of ob-session-async mode jackkamm/ob-session-async#7.
Here is how you may be able to reproduce it (it looks long but I've tried my best to make it minimal):
```elisp
(defun test-macos-hang ()
  (require 'ess-inf)
  (with-temp-buffer
    (insert "Sys.sleep(5)")
    (let ((ess-dialect "R"))
      (ess-eval-buffer nil)))
  (with-temp-buffer
    (insert
     "{
    function(object,transfer.file) {
        object
        invisible(
            if (
                inherits(
                    try(
                        {
                            tfile<-tempfile()
                            write.table(object,
                                        file=tfile,
                                        row.names = FALSE,
                                        col.names = FALSE,
                                        quote=FALSE)
                            file.rename(tfile,transfer.file)
                        },
                        silent=TRUE),
                    \"try-error\"))
                {
                    if(!file.exists(transfer.file))
                        file.create(transfer.file)
                }
            )
    }
}(object=\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\",transfer.file=tempfile())"
     )
    (let ((ess-dialect "R"))
      (ess-eval-buffer nil))))
```
Evaluate the script and call the function `test-macos-hang` inside Emacs, Emacs will hang until the evaluation is finished, which should be around 5 seconds because of `Sys.sleep(5)`.

What I expected to see was all the R evaluation should be done in the background and I could continue to use Emacs for other tasks. And this is actually the case when I test under Windows 10 (haven't tested under Linux)

One thing to mention is if I reduce the number of `a`'s for the object string near the end of the script, or remove some line breaks (/n) after the parentheses/brackets, then the hanging will not happen. Could it be that Emacs's poor performance dealing with long strings caused this issue?

**Comment by lionel-** [comment]
This is a duplicate of #1093 right?

**Comment by ywwry66** [comment]
They don't seem to be the same to me, but could be related. I found #1093 when I was playing with this. That one can be reduced to a neat minimal example, so I filed it first (and I was hoping if that one gets resolved, maybe this will go away as well).

The `Sys.sleep()` part is just used to mimic a time consuming R procedure. If it is removed, the evaluation is done instantly, and the hang won't be noticed even if it is produced. In case you didn't see, the hang happens precisely when the R code gets evaluated.

I have edited the example to kill the side effect. If you can't reproduce, maybe try adding a couple extra `a`'s to the object string near the bottom. If you ever use org-mode, could you test if the package https://github.com/jackkamm/ob-session-async would work on macOS with the following toy example? This is where these investigations all begin
```org
#+begin_src R :async :session
  Sys.sleep(5)
  "this won't hang your emacs"
#+end_src
```

**Comment by lionel-** [comment]
Oh so you mean ESS hangs during the 5 seconds of the `Sys.sleep()` call?

What is the role of the second snippet of R code in your example?

**Comment by lionel-** [comment]
I'm starting to see more clearly. Yes I can reproduce the 5 seconds hang of your example. These two examples don't hang on the other hand, suggesting Emacs blocks when sending long inputs:

```elisp
(defun test-macos-hang2 ()
  (require 'ess-inf)
  (with-temp-buffer
    (insert "Sys.sleep(5)")
    (let ((ess-dialect "R"))
      (ess-eval-buffer nil))))

(test-macos-hang2)


(defun test-macos-hang3 ()
  (require 'ess-inf)
  (with-temp-buffer
    (insert "Sys.sleep(5)")
    (let ((ess-dialect "R"))
      (ess-eval-buffer nil)))
  (with-temp-buffer
    (insert "Sys.sleep(2)")
    (let ((ess-dialect "R"))
      (ess-eval-buffer nil))))

(test-macos-hang3)
```

Indeed this one hangs:

```lisp
(defun test-macos-hang4 ()
  (require 'ess-inf)
  (with-temp-buffer
    (insert "Sys.sleep(5)")
    (let ((ess-dialect "R"))
      (ess-eval-buffer nil)))
  (with-temp-buffer
    (insert "{
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            }")
    (let ((ess-dialect "R"))
      (ess-eval-buffer nil))))

(test-macos-hang4)
```

Interrupting shows `^G` sequences in the REPL as in #1093.

**Comment by lionel-** [comment]
Let's track this in #1093. Sorry but this is going to be low priority. If you'd like to investigate what's happening that'd be great! This is potentially a deep rabbit hole unfortunately.

**Comment by ywwry66** [comment]
Right. Because normally when `ess-eval-buffer` is called, R just runs in the background and I can continue to do other work inside Emacs.

The second snippet comes from the variable `org-babel-R-write-object-command` defined in `org-babel`, and it is used to write `object` in `transfer.file`. And the hang won't be produced without that (probably just because of its length)

**Comment by ywwry66** [comment]
Of course I'd like to investigate, but I may lack the programming skill. Anyways, I will keep you post if I find anything.

**Comment by lionel-** [comment]
oh if I disable tracebug it seems this fixes the hang!

This might unblock you. I don't think you need tracebug enabled for literate programming.

**Comment by ywwry66** [comment]
This does fix the hang with my example, but seems to fail for other longer examples. Just try adding several rows of semicolons to your `test-macos-hang4`. So yes, it is a deep rabbit hole I would say.

**Comment by ywwry66** [comment]
Hi @lionel- , can you elaborate on your comment "Interrupting shows ^G sequences in the REPL" about `test-macos-hang4`? I can't reproduce this when I use `C-c C-c` to interrupt.

I am testing an Emacs patch which solves #1093, but `test-macos-hang4` still causes hang. So I was wondering if `^G` can be observed with the patch.
